test/UnbrandedProjects.Tests/ClientUriBuilderTests.cs

```diff
@@ -0,0 +1,461 @@
+        [TestCase("http://localhost/with/", "/path", "http://localhost/with/path")]
+        [TestCase("http://localhost:12345/with", "/path", "http://localhost:12345/with/path")]
+        [TestCase("http://localhost:12345/with/", "/path", "http://localhost:12345/with/path")]
+        [TestCase("http://localhost", "/path/", "http://localhost/path/")]
+        [TestCase("http://localhost/", "/path/", "http://localhost/path/")]
+        [TestCase("http://localhost:12345", "/path/", "http://localhost:12345/path/")]
+        [TestCase("http://localhost:12345/", "/path/", "http://localhost:12345/path/")]
+        [TestCase("http://localhost/with", "/path/", "http://localhost/with/path/")]
+        [TestCase("http://localhost/with/", "/path/", "http://localhost/with/path/")]
+        [TestCase("http://localhost:12345/with", "/path/", "http://localhost:12345/with/path/")]
+        [TestCase("http://localhost:12345/with/", "/path/", "http://localhost:12345/with/path/")]
+        public void ResetThenAppendPath(string endpoint, string pathPart, string expected)
+        {
+            var builder = new ClientUriBuilder();
+            builder.Reset(new Uri(endpoint));
+
+            if (pathPart != null)
+            {
+                builder.AppendPath(pathPart, false);
+            }
+
+            Assert.AreEqual(expected, builder.ToUri().ToString());
+        }
+
+        [TestCase("http://localhost", "one|1", "http://localhost/?one=1")]
+        [TestCase("http://localhost:12345", "one|1", "http://localhost:12345/?one=1")]
+        [TestCase("http://localhost/some", "one|1", "http://localhost/some?one=1")]
+        [TestCase("http://localhost:12345/some", "one|1", "http://localhost:12345/some?one=1")]
+        [TestCase("http://localhost/", "one|1", "http://localhost/?one=1")]
+        [TestCase("http://localhost:12345/", "one|1", "http://localhost:12345/?one=1")]
+        [TestCase("http://localhost/some/", "one|1", "http://localhost/some/?one=1")]
+        [TestCase("http://localhost:12345/some/", "one|1", "http://localhost:12345/some/?one=1")]
+        [TestCase("http://localhost", "one|1|two|2", "http://localhost/?one=1&two=2")]
+        [TestCase("http://localhost:12345", "one|1|two|2", "http://localhost:12345/?one=1&two=2")]
+        [TestCase("http://localhost/some", "one|1|two|2", "http://localhost/some?one=1&two=2")]
+        [TestCase("http://localhost:12345/some", "one|1|two|2", "http://localhost:12345/some?one=1&two=2")]
+        [TestCase("http://localhost/", "one|1|two|2", "http://localhost/?one=1&two=2")]
+        [TestCase("http://localhost:12345/", "one|1|two|2", "http://localhost:12345/?one=1&two=2")]
+        [TestCase("http://localhost/some/", "one|1|two|2", "http://localhost/some/?one=1&two=2")]
+        [TestCase("http://localhost:12345/some/", "one|1|two|2", "http://localhost:12345/some/?one=1&two=2")]
+        public void ResetThenAppendQuery(string endpoint, string queryPart, string expected)
```

Comment:
Would we ever want a case where we append a query param where the parameter name has a leading '?' and the test validates that our type removes the leading query (i.e. only one '?' appears in the resulting URI) ?

Reply:
yeah we may need those new test cases, and I am deferring your PR to consolidate those cases: https://github.com/Azure/azure-sdk-for-net/pull/43436